doc/changelog.md

```diff
@@ -15,6 +15,9 @@
   [(#895)](https://github.com/PennyLaneAI/catalyst/pull/895)
 
 <h3>Bug fixes</h3>
+* Using float32 in callback functions would not crash in compilation phase anymore,
+  but rather raise an assertion error.
+  [(#916)]https://github.com/PennyLaneAI/catalyst/pull/916
```

Comment:
Is there no way to fix this for real? Or are there some deeper reasons we can't use f32

Reply:
Originally, we decided against using f32 due to the runtime only supporting `double` precision. Of course that is only for quantum instructions, so there's no need to extend this constraint to the classical side. I suspect though, that one of the reasons we see this is the fact that we override JAX's defaults to use f64. I think a potential fix here would be to add a cast. This could be bad for performance, but good for user experience. Consider really large tensors of f64 that need to be converted to f32 precision.

```python
@pure_callback
def foo() -> (ShapedArray([], float32):
  # dtype is float64 by default
  # due to this line: jax.config.update("jax_enable_x64", True)
  return jax.numpy.array(0.0, dtype=float64)
  # only solution is to fix the line above or cast to float32
```

Reply:
@mehrdad2m can you try a test case where the callback returns float32 and the dtype in the array is set to float32 as well?

Reply:
> I suspect though, that one of the reasons we see this is the fact that we override JAX's defaults to use f64.

Yes we do enable JAX f64 support, but that doesn't mean all floats are suddenly 64 bits. Python floats will be interpreted as 64 (most of the time, unless they interact with a more specific dtype, in which case they will conform to the latter), but all explicit dtypes should remain unchanged. So in theory, there should be no issue with the user explicitly creating a 32bit float and passing it to and from a function 🤔

Reply:
Actually yes it would be need nice to include some test cases where this error actually arises :)

Reply:
> Yes we do enable JAX f64 support, but that doesn't mean all floats are suddenly 64 bits.
> So in theory, there should be no issue with the user explicitly creating a 32bit float and passing it to and from a function 🤔

I think we agree here, but for some reason I feel like I have to clarify this a bit further. I think the error is that the value returned from the pure_callback was not `dtype`d float32, which is why the assertion is raised. The test was this:

```
@pure_callback
def foo() -> ShapedArray([], float32):
  return jax.numpy.array(0.0) # which is implicitly dtyped float64
# at run time the type of the returned value is checked against ShapedArray([], float32)
# if they don't match, an assertion is raised. (I think)
```

Reply:
Added two test cases that show case this issue. basically the user should explicitly set the type to float32 for both input and output, which would result in normal execution of the code. if output type is not set, the assertion error will be raised.

Reply:
> loat32 for both input and output,

Just a small correction, one is the expected output type and the other is the observed output type. If they don't match exactly at the moment, the behaviour is that an error is raised. I think this fixes the issue completely and if we want to automatically cast that is a different question.

Thanks @mehrdad2m !

Reply:
Updated the PR message so it wouldn't cause confusion later.

Reply:
> > loat32 for both input and output,
> 
> Just a small correction, one is the expected output type and the other is the observed output type. If they don't match exactly at the moment, the behaviour is that an error is raised. I think this fixes the issue completely and if we want to automatically cast that is a different question.
> 
> Thanks @mehrdad2m !

Oh sorry you are right. my bad!

Reply:
@dime10 added the error messages to PR message.